Mathlib/RingTheory/Radical.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2024 Jineon Back and Seewoo Lee. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jineon Baek, Seewoo Lee
+-/
+import Mathlib.RingTheory.UniqueFactorizationDomain
+
+/-!
+# Radical of an element of a unique factorization normalization monoid
+
+This file defines a radical `rad(a)` of an element `a` of a unique factorization normalization
+monoid, which is defined as a product of normalized prime factors of `a` without duplication.
+This is different from the radical of ideal.
+
+## Main declarations
+
+- `radical_associated_eq`: If `a` and `b` are associates, i.e. `a * u = b` for some unit `u`,
+  then `rad(a) = rad(b)`.
+- `radical_unit_hMul`: Multiplying unit does not change the radical.
+- `radical_dvd_self`: `rad(a)` divides `a`.
+- `radical_pow`: `rad(a ^ n) = rad(a)` for any `n ≥ 1`
+- `radical_prime`: Radical of a prime element is equal to its normalization
+
+## TODO
+
+- Make a comparison with `Ideal.radical`. Especially, for principal ideal, `rad((a)) = (rad(a))`,
+  where the first (resp. second) `rad` is radical of an ideal (resp. element).
+- Prove `rad(rad(a)) = rad(a)`.
+-/
+
+noncomputable section
+
+open scoped Classical
+
+open UniqueFactorizationMonoid
+
+variable {k : Type _} [Field k]
+variable {α : Type _} [CancelCommMonoidWithZero α] [NormalizationMonoid α]
```

Comment:
Do you need the `0` element for the definition? Or is an arbitrary monoid good enough?
Also, please call the monoid `M`.

Reply:
No mathematically, but it's that [`UniqueFactorizationMonoid`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/UniqueFactorizationDomain.html#UniqueFactorizationMonoid) requires a `CancelCommMonoidWithZero` by the definition in Mathlib, and we need unique factorization to talk about primes.